src/EnergyPlus/ChillerElectricASHRAE205.cc

```diff
@@ -195,8 +195,8 @@ void getChillerASHRAE205Input(EnergyPlusData &state)
         //            ErrorsFound = true;
         //        }
 
-        auto const evap_inlet_node_name = ip->getAlphaFieldValue(fields, objectSchemaProps, "chilled_water_inlet_node_name");
-        auto const evap_outlet_node_name = ip->getAlphaFieldValue(fields, objectSchemaProps, "chilled_water_outlet_node_name");
+        std::string const &evap_inlet_node_name = ip->getAlphaFieldValue(fields, objectSchemaProps, "chilled_water_inlet_node_name");
+        std::string const &evap_outlet_node_name = ip->getAlphaFieldValue(fields, objectSchemaProps, "chilled_water_outlet_node_name");
```

Comment:
There is all sorts of string copying happening within `InputProcessor` in the `getAlphaFieldValue` call, so I don't think this `evap_outlet_node_name` is pointing to a persistent object here. The safe and fast thing to do is probably `std::string const evap_outlet_node_name = std::move(ip->getAlphaFieldValue(fields, objectSchemaProps, "chilled_water_outlet_node_name");`. But the better thing to do would be to look at the `InputProcessor` implementation and clean it up.

Reply:
getAlphaFieldValue returns a std::string. That string should just be assigned to a local string. As far as the std::move, I am not familiar with what that would do. It compiles and runs, I just don't know enough to use it wisely.

    std::string InputProcessor::getAlphaFieldValue() {
        std::string value;
        value = find the string;
        return value;
    }

Reply:
A std::string is a two-part object. There is the part that contains the actual characters. This part lives in a section of memory called the "heap" and is dynamically allocated, deallocated, and can be resized. Then there is the header/wrapper that contains a pointer to the actual string plus some other information, e.g., the amount of memory currently allocated to the character string, maybe the length of the current string in that memory. This part lives in the "enclosing context". If the `std::string` is declared as a local variable, this header is essentially part of the stack frame. If the `std::string` is a member in an object, this header is a contiguous part of the object, it's essentially embedded in the object.

When you are doing a normal string copy, i.e., `std::string s1 = s2;` you are doing two things. First you are copying the header of `s2` into the header of `s1`. This is usually fast because the header is small. Then you are allocating another chunk of heap memory for `s1` to point to and copying the contents of the character string pointed to by `s2` into it. So now you have two copies of the header *and* two copies of the character string. Essentially, you've made a "deep copy" of the string.

It is not possible to avoid copying the header, but it is possible to avoid making a deep copy of the character string on the heap. You do it like this: `std::string s1 = std::move(s2);`. This activates the `std::string` "move constructor" which copies the `s2` header into `s1`, including the pointer to the heap memory pointed to by `s2`. Essentially, `s2` is "handing off" the character string in heap memory to `s1`. Now you have two copies of the header, but only one copy of the character string in the heap. You've done a shallow copy.

The question is when do you want to do this as opposed to making a deep copy. Well, you want to do it when you are sure that `s2` will not be accessed anymore.  One example of that is when `s2` goes out of scope. And one example of that is when `s2` is created in and then returned by a function. For instance, you can do `std::string s1 = std::move(getAlphaFieldValue());` safely because the return value of `getAlphaFieldValue()` is out of scope after this statement, i.e., there is no way to access it anymore.

Reply:
I added this explanation to the EnergyPlus wiki for future reference.

Reply:
@Myoldmopar @amirroth I think this is the conversation and area of code that used std::move

Reply:
Thx. So, `std::move` predates RVO/copy-elision by a little bit. RVO is more powerful (`std::move` is a "shallow copy" and RVO is a "no copy") for return values while `std::move` can be used in general copy situations. And while the two don't conflict logically speaking (I don't understand why a compiler cannot elide a "move" just as easily as it can elide a "copy"), maybe there are some corner cases where an elided `std::move` would behave differently than a non-elided one (I cannot think of an example off the top of my head, but that doesn't mean one doesn't exist) and clang has a policy to keep `std::move` usage "backwards compatible". According to [this SO post](https://stackoverflow.com/questions/17473753/c11-return-value-optimization-or-move) (which is actually seven years old!) you should not `std::move` return values and rely on the compiler to RVO them. I don't know that this is universally good advice (it's difficult to RVO large objects), but it's probably good enough for strings. Live and learn.